lib/engine-addon.js

```diff
@@ -330,36 +333,27 @@ module.exports = {
       let invokeArguments = args || [];
       let hostAddons = this.ancestorHostAddons();
 
-      return this.addons
-        .map(addon => {
+      if (methodName === 'treeFor') {
+        deeplyNonDuplicatedAddon(hostAddons, this, invokeArguments[0]);
+      }
 
+      let trees = this.addons
+        .filter(addon => {
           if (!addon[methodName]) {
             // no method to call
-            return;
+            return false;
           }
+          return true;
+        })
+        .map(addon => {
+          return addon[methodName].apply(addon, invokeArguments);
+        });
 
-          let hostAddon = hostAddons[addon.name];
-
-          if (hostAddon && hostAddon.cacheKeyForTree) {
-            if (methodName === 'treeFor') {
-              let treeName = invokeArguments[0];
-              let hostAddonCacheKey = hostAddon.cacheKeyForTree(treeName);
-              let addonCacheKey = addon.cacheKeyForTree(treeName);
-
-              if (
-                addonCacheKey != null &&
-                addonCacheKey === hostAddonCacheKey
-              ) {
-                // the addon specifies cache key and it is the same as host instance of the addon, skip the tree
-                return;
-              }
-            } else {
-              return;
-            }
-          }
+      if (methodName === 'treeFor') {
+        recoverAddons(this);
```

Comment:
As we are doing some semi dangerous monkey-patching here, the restoration should occur within a `finally` clause.

Reply:
You mean the `Promise.prototype.finally` or some `finally` hook?

Reply:
```
try {
  // monkey patch here
} finally {
  // restore here
}
```

Reply:
Get it, thank you @rwjblue .